grails-app/controllers/org/pih/warehouse/inventory/InventoryController.groovy

```diff
@@ -638,52 +598,30 @@ class InventoryController {
     }
 
     def listQuantityOnHandZero = {
-        def location = Location.get(session.warehouse.id)
-        def inventoryItems = dashboardService.getQuantityOnHandZero(location)
-
-        if (params.format == "csv") {
-            def filename = "Out of stock  - all - " + location.name + ".csv"
-            response.setHeader("Content-disposition", "attachment; filename=\"${filename}\"")
-            render(contentType: "text/csv", text: getCsvForProductMap(inventoryItems))
-            return
-        }
-
-        render(view: "list", model: [availableItems: inventoryItems])
+        this.listStock(params, "getQuantityOnHandZero", "Out of stock  - all - ")
     }
 
     def listHealthyStock = {
-        def location = Location.get(session.warehouse.id)
-        def inventoryItems = dashboardService.getHealthyStock(location)
-
-        if (params.format == "csv") {
-            def filename = "Overstock - " + location.name + ".csv"
-            response.setHeader("Content-disposition", "attachment; filename=\"${filename}\"")
-            render(contentType: "text/csv", text: getCsvForProductMap(inventoryItems))
-            return
-        }
-
-        render(view: "list", model: [availableItems: inventoryItems])
+        this.listStock(params, "getHealthyStock", "Overstock - ")
     }
 
     def listOverStock = {
-        def location = Location.get(session.warehouse.id)
-        def inventoryItems = dashboardService.getOverStock(location)
-
-        if (params.format == "csv") {
-            def filename = "Overstock - " + location.name + ".csv"
-            response.setHeader("Content-disposition", "attachment; filename=\"${filename}\"")
-            render(contentType: "text/csv", text: getCsvForProductMap(inventoryItems))
-            return
-        }
-
-        render(view: "list", model: [availableItems: inventoryItems])
+        this.listStock(params, "getOverStock", "Overstock - ")
     }
 
     def listOutOfStock = {
         def location = Location.get(session.warehouse.id)
-        def inventoryItems = dashboardService.getOutOfStock(location, params.abcClass)
+        List<Category> categories = Category.findAllByIdInList(params.list("categories"))
 
-        if (params.format == "csv") {
+        // When accessing the page for the first time, the flag should be set to true
+        // Initially there not parameter button, only after running the report manually it is set
+        params.includeSubcategories = params.button ? params.includeSubcategories : true
+        if (params.includeSubcategories) {
+            categories = inventoryService.getExplodedCategories(categories)
+        }
+        def inventoryItems = dashboardService.getOutOfStock(location, params.abcClass, categories)
```

Comment:
Initially I wanted to write if you forgot to use the re-usable logic in this method as well, but I can see that it's a bit different than the others.
How about creating another method called, let's say `determineCategories` (ofc the name doesn't have to be like this) and doing there (pseudocode written by hand)
```groovy
List<Category> determineCategories(params) {
  List<Category> categories = Category.findAllByIdInList(params.list("categories"))
  params.includeSubcategories = params.button ? params.includeSubcategories : true
  if (params.includeSubcategories) {
    categories = inventoryService.getExplodedCategories(categories)
  }
}
```
and use it here and also change the `listStock` 488th line method to:
```groovy
...
List<Category> categories = determineCategories(params)
...
```